content/blog/2022-04-04-develop-and-test-data-analytics-pipelines-on-localstack/index.md

```diff
@@ -0,0 +1,236 @@
+
+## Defining the Data Pipeline
+At LocalStack we use AWS [Cloud Development Kit](https://aws.amazon.com/cdk/) (CDK) to define our infrastructure as code. We can use the [cdklocal](https://github.com/localstack/aws-cdk-local) tool to deploy the same CDK stacks on our development machines under LocalStack as we do in production. This allows us to test our CDK code, and therefore our infrastructure itself, without needing to deploy.
+Okay, let's define a CDK stack for our data pipeline! We'll start with a boilerplate class definition:
+
+```python
+# data_pipeline_stack.py
+
+
+class DataPipelineStack(Stack):
+    def __init__(
+        self,
+        scope: Construct,
+        construct_id: str,
+        monitored_log_group_arn: str,
+        tinybird_auth_token: str,
+        tinybird_url: str,
+        **kwargs
+    ) -> None:
+        super().__init__(scope, construct_id, **kwargs)
+```
+
+If you're familiar with CDK there shouldn't be anything surprising here. We declare a `DataPipelineStack` as a CDK Stack class. We also define a few parameters in the constructor that will be useful to us later: the ARN of the CloudWatch log group associated with the Lambda we want to monitor, plus an auth token and API URL for communicating with our Tinybird data warehouse.
+Now let's define the beginning of the pipeline.
+
+```python
+# data_pipeline_stack.py
+
+
+kinesis_stream = aws_kinesis.Stream(
+    self,
+    "data_pipeline_kinesis_stream",
+    shard_count=1,
+)
+
+monitored_log_group = aws_logs.LogGroup.from_log_group_arn(
+    self, "monitored_log_group", monitored_log_group_arn
+)
+monitored_log_group.add_subscription_filter(
+    "monitored_log_group_cloudwatch_subscription",
+    destination=aws_logs_destinations.KinesisDestination(kinesis_stream),
+    filter_pattern=aws_logs.FilterPattern.all(
+        aws_logs.FilterPattern.exists("$.event_class"),
+        aws_logs.FilterPattern.exists("$.event_type"),
+    ),
+)
+
+```
+First we stand up a Kinesis stream for our event data. Next, we import the log group associated with the Lambda function we're interested in using its ARN. Finally, we declare a subscription filter using these two elements. The subscription filter acts as the "glue" between CloudWatch and Kinesis. Note the filter pattern: this ensures that only the structured events that we care about will be sent to Kinesis. We don't want every Lambda log message to show up in our data warehouse! Our filter only accepts log lines that meet the following criteria:
+* JSON encoded
+* Contains an `event_class` field
+* Contains an `event_type` field
+
+This filter was chosen based on the particular analytics event format we use at LocalStack.
```

Comment:
It redirects to a private repo link. Am I understanding right that this is intended here?